### PR TITLE
docs: plan review runbooks design

### PR DESCRIPTION
**Pull request by tobyhede**
Three-stage pipeline (gate → parallel dual-review → synthesis) that decomposes the 35-item plan review checklist into focused runbooks for parallel agent execution. Identifies two infrastructure prereqs: FOR + step-level runbooks support, and implicit loop aggregation fix.

## Summary by CodeRabbit

* **Tests**
  * Greatly expanded coverage for parent-state aggregation, complex loop patterns, iteration tracking, GOTO/NEXT/RETRY behaviors, last-action accuracy, and multi-step edge cases.

* **Refactor**
  * Overhauled aggregation and transition routing in the runbook compiler to improve iteration/result accumulation and transition correctness.
  * Runtime context now tracks parent-level retry counts to support parent-aggregation flows.

### CODE REVIEW

File: packages/core/src/runbook/compiler.ts [hunk 1]
```diff
@@ -571,6 +538,153 @@ function buildAggregationExitAssign(
   }
 }
 
+/**
+ * Build the `always` (eventless) transition configuration for a parent aggregation state.
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts]: Verified: the word "eventless" does not appear in the current code. Already resolved in a prior commit.

File: packages/core/src/runbook/compiler.ts [hunk 2]
```diff
@@ -1009,8 +1034,9 @@ export function compileRunbookToMachine(
         }
       : {};
 
-    // Build per-state GOTO transitions
-    const buildGotoTransitionsForState = allStates.map((target) => {
+    // Build per-state GOTO transitions (skip parent states — they are transient)
+    const gotoTargets = allStates.filter((t) => !t.isParentState);
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts]: Fixed in 427ea29b — `gotoTargets` is now computed once before the `allStates.forEach` loop.

File: packages/core/src/runbook/compiler.ts [hunk 3]
```diff
@@ -68,6 +68,19 @@ interface TransitionEntry {
...
+ * Parent aggregation states (isParentState=true) are built via buildParentStateConfig.
+ */
+interface StateConfig {
+  id: string;
+  stepName: string;
+  substepId?: string;
+  transitions: Transitions;
+  isParentState?: boolean;
+  parentStep?: Step;
+}
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts]: Fixed in 427ea29b — replaced with discriminated union (`ChildStateConfig | ParentStateConfig`). `buildParentStateConfig` now accepts `ParentStateConfig` directly; redundant runtime guard removed.

File: packages/core/src/runbook/compiler.ts [hunk 4]
```diff
@@ -571,6 +538,153 @@ function buildAggregationExitAssign(
...
+        (transition.retry <= 0 || context.retryCount >= transition.retry),
+      target,
+      actions: [
+        buildParentExitAssign(transition.action, target, steps, sources),
+        assign({
+          retryCount: 0,
+          retryMax: transition.retry > 0 ? transition.retry : undefined,
+        }),
+      ],
+    };
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:683]: Fixed in 427ea29b — removed redundant `retryCount: 0` and `parentRetryCount: 0` from the second assign. Only `retryMax` remains.

File: packages/core/src/runbook/compiler.ts [hunk 5]
```diff
@@ -571,6 +538,153 @@ function buildAggregationExitAssign(
...
+          substep: firstSubstep?.id,
+        }),
+      },
+      exhausted,
+    ];
+  };
+
+  // Aggregation guards (Cases A & B: steps with explicit transitions)
+  if (hasTransitions) {
+    const parentTransitions = parentStep.transitions!;
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts]: Verified: `parentStep.transitions\!` (with `\!`) does not appear in the current code. Already resolved in prior commits.

File: packages/core/__tests__/runbook/compiler.test.ts [hunk 6]
```diff
@@ -4111,4 +4219,593 @@ echo "processing"
...
+
+      actor.send({ type: 'FAIL' }); // 1.2 -> parent -> retry #1
+      expect(actor.getSnapshot().value).toBe('step::1::1');
+
+      actor.send({ type: 'FAIL' }); // 1.1 -> 1.2
+      expect(actor.getSnapshot().value).toBe('step::1::2');
+
+      actor.send({ type: 'FAIL' }); // 1.2 -> parent -> retries exhausted
+      expect(actor.getSnapshot().value).toBe('STOPPED');
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/compiler.test.ts:4774]: Verified: already addressed in commit 107f726.

File: packages/core/src/runbook/compiler.ts [hunk 7]
```diff
@@ -571,6 +550,157 @@ function buildAggregationExitAssign(
...
+      actions: assign({
+        forStack: ({ context }: { context: RunbookContext }) => {
+          const top = peekForStack(context.forStack);
+          if (!top) return context.forStack;
+          return [{ ...top, iteration: nextIteration(top), currentValue: undefined }];
+        },
+        retryCount: 0,
+        substep: firstSubstep?.id,
+      }),
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:661]: Won't fix — `parentRetryCount` intentionally persists across loop iterations. It tracks cumulative parent-level retry attempts; resetting at loop-back would break retry exhaustion detection. The scenario described (iteration N retry leaking to N+1) is actually correct behavior: `parentRetryCount` tracks how many times the *parent aggregation* has retried the entire loop, not per-iteration state. It is correctly reset only on parent state exit via `buildParentExitAssign`/`baseAssign`.

File: packages/core/src/runbook/compiler.ts [hunk 8]
```diff
@@ -367,6 +373,23 @@ function buildSimpleGotoAssign(options: {
...
+
+function buildLoopControlAssign(actionType: 'NEXT' | 'BREAK', iterationResult: 'pass' | 'fail') {
+  return {
+    iterationResults: appendIterationResult(iterationResult),
+    lastAction: { type: actionType },
+    lastMessage: undefined as string | undefined,
+    retryCount: 0,
+    substep: undefined as string | undefined,
+  };
+}
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts]: Verified: `buildLoopControlAssign` does not exist in the current code. Already resolved in a prior commit.

File: packages/core/src/runbook/compiler.ts [hunk 9]
```diff
@@ -367,6 +373,36 @@ function buildSimpleGotoAssign(options: {
   });
 }
 
+function appendIterationResult(iterationResult: 'pass' | 'fail') {
+  return ({ context }: { context: RunbookContext }): ('pass' | 'fail')[] => {
+    const results = context.iterationResults ?? [];
+    return [...results, iterationResult];
+  };
+}
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:399]: Won't fix — ESLint does not flag this function. Ran `npx eslint packages/core/src/runbook/compiler.ts` locally with zero errors. The `allowExpressions: true` config exempts it (the returned arrow function expression has an inline return type annotation).

File: packages/core/src/runbook/compiler.ts [hunk 10]
```diff
@@ -571,6 +584,159 @@ function buildAggregationExitAssign(
...
+          retryCount: ({ context }: { context: RunbookContext }) => context.retryCount + 1,
+          retryMax: transition.retry,
+          forStack: [] as readonly ForContext[],
+          iterationResults: [] as ('pass' | 'fail')[],
+          lastMessage: undefined as string | undefined,
+          substep: firstSubstep?.id,
+        }),
+      },
+      exhausted,
+    ];
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:707]: Won't fix — the existing comment at lines 696-697 already documents the intent: "Increment both counters: parentRetryCount tracks parent-level exhaustion, retryCount is exposed to the execution layer (actor-service) for visibility." This is option (B) from the suggestion, already implemented.

File: packages/core/src/runbook/compiler.ts [hunk 11]
```diff
@@ -571,6 +584,159 @@ function buildAggregationExitAssign(
...
+      return shouldAggregationPass(hasFailed, passCount, parentTransitions.all);
+    };
+
+    const passBranchGuard: GuardFn = aggregationPasses;
+    const failBranchGuard: GuardFn = ({ context }) => !aggregationPasses({ context });
+
+    always.push(
+      ...buildOutcomeEntries(passBranchGuard, parentTransitions.pass, passTarget),
+      ...buildOutcomeEntries(failBranchGuard, parentTransitions.fail, failTarget),
+    );
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:729]: Won't fix — these are XState `always` transition guards evaluated by the state machine runtime. Each guard is called independently; you cannot cache across invocations. In the pass case, `aggregationPasses` is only evaluated once (XState takes the first matching transition). In the fail case it is evaluated twice, but restructuring to avoid this (e.g., unguarded fallback) would be fragile and save negligible cost on a linear array scan.

File: packages/core/src/runbook/compiler.ts [hunk 12]
```diff
@@ -571,6 +584,159 @@ function buildAggregationExitAssign(
...
+    if (!hasFor) {
+      // Case D: non-FOR pass-through — clear iterationResults, set CONTINUE
+      exitAssign.iterationResults = undefined as ('pass' | 'fail')[] | undefined;
+      exitAssign.lastAction = { type: 'CONTINUE' as const };
+      exitAssign.lastMessage = undefined as string | undefined;
+    }
+    // Case C: FOR without transitions — preserve lastAction from substep
+
+    always.push({ target: nextTarget, actions: assign(exitAssign) });
+  }
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/compiler.ts:748]: Won't fix — the current behavior is correct by design. In XState `assign()`, fields not present in the object are preserved from context. Case C intentionally omits `lastAction` so the substep's value passes through unchanged. Adding `exitAssign.lastAction = ({ context }) => context.lastAction` would be a functional no-op. The comment at line 745 documents this.
